kv/kvserver: TestLogGrowthWhenRefreshingPendingCommands failed

kv/kvserver.TestLogGrowthWhenRefreshingPendingCommands [failed](https://mesolite.cluster.engflow.com/invocations/default/76a75d9e-4776-4c3b-9c21-266d75fd0fda?testReportRun=1&testReportShard=14&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [ac68bcaf4a547b9d24127a6da8ee01df7753fea1](https://github.com/cockroachdb/cockroach/commits/ac68bcaf4a547b9d24127a6da8ee01df7753fea1):


```
=== RUN   TestLogGrowthWhenRefreshingPendingCommands
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestLogGrowthWhenRefreshingPendingCommands262996019
    test_log_scope.go:82: use -show-logs to present logs inline
    client_raft_test.go:2032: -- test log scope end --
test logs left over in: outputs.zip/logTestLogGrowthWhenRefreshingPendingCommands262996019
--- FAIL: TestLogGrowthWhenRefreshingPendingCommands (153.58s)
=== RUN   TestLogGrowthWhenRefreshingPendingCommands/lease-type=LeaseLeader
--- FAIL: TestLogGrowthWhenRefreshingPendingCommands/lease-type=LeaseLeader (58.31s)
=== RUN   TestLogGrowthWhenRefreshingPendingCommands/lease-type=LeaseLeader/proposeOnFollower=false
    test_server_shim.go:561: DRPC is enabled (override by TestingGlobalDRPCOption)
    testcluster.go:1223: lease is now of type: LeaseLeader
    client_raft_test.go:2020: write finished with quorum unavailable; err=[NotLeaseHolderError] refusing to acquire lease on follower; r79: replica (n1,s1):1 not lease holder; lease holder unknown
--- FAIL: TestLogGrowthWhenRefreshingPendingCommands/lease-type=LeaseLeader/proposeOnFollower=false (27.29s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=14</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogGrowthWhenRefreshingPendingCommands.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-54875

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -1878,6 +1878,10 @@ func TestLogGrowthWhenRefreshingPendingCommands(t *testing.T) {
 			raftConfig := base.RaftConfig{
 				// Drop the raft tick interval so the Raft group is ticked more.
 				RaftTickInterval: 10 * time.Millisecond,
+				// Suppress timeout-based elections to avoid leadership changes in ways
+				// this test doesn't expect. See DisablePreCampaignStoreLivenessCheck,
+				// which counteracts this knob's effect on leader leases.
+				RaftElectionTimeoutTicks: 1000000,
 				// Reduce the max uncommitted entry size.
 				RaftMaxUncommittedEntriesSize: 64 << 10, // 64 KB
 				// RaftProposalQuota cannot exceed RaftMaxUncommittedEntriesSize.
@@ -1885,12 +1889,6 @@ func TestLogGrowthWhenRefreshingPendingCommands(t *testing.T) {
 				// RaftMaxInflightMsgs * RaftMaxSizePerMsg cannot exceed RaftProposalQuota.
 				RaftMaxInflightMsgs: 16,
 				RaftMaxSizePerMsg:   1 << 10, // 1 KB
-			}
-			// Suppress timeout-based elections to avoid leadership changes in ways this
-			// test doesn't expect. For leader leases, fortification itself provides us
-			// this guarantee.
-			if leaseType != roachpb.LeaseLeader {
-				raftConfig.RaftElectionTimeoutTicks = 1000000
 			}
 
 			const numServers int = 5
@@ -1916,6 +1914,14 @@ func TestLogGrowthWhenRefreshingPendingCommands(t *testing.T) {
 							// Refresh pending commands on every Raft group tick instead of
 							// every RaftReproposalTimeoutTicks.
 							RefreshReasonTicksPeriod: 1,
+							RaftTestingKnobs: &raft.TestingKnobs{
+								// Due to high RaftElectionTimeoutTicks, we only have one
+								// opportunity to campaign which should not be missed. Under
+								// leader leases, in a "cold" cluster, a campaign can fail due
+								// to missing store liveness support from the node's peers.
+								// Disallow this check to ensure that the campaign succeeds.
+								DisablePreCampaignStoreLivenessCheck: true,
+							},
 						},
 					},
 				}

